### PR TITLE
Call Ipopt solver explicitly when solving static equilibrium problem.

### PR DESCRIPTION
**Pull request by hongkai-dai**
Resolves a weekly build CI failure.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14199)

### CODE REVIEW

**Comment by hongkai-dai** [comment]
@drake-jenkins-bot  linux-focal-gcc-bazel-weekly-gurobi-debug please.

**Review by SeanCurtis-TRI — COMMENTED** [review]
Could you elaborate on why this is necessary?

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14199#-:-MJSkHU_0cGZD2rBvDST:b-8coou9)* status: needs platform reviewer assigned, needs at least two assigned reviewers

**Comment by hongkai-dai** [comment]
Previously when snopt is enabled, the optimization program will call snopt solver, and it timed out on Ubuntu focal. I just switched to Ipopt to check the performance.

How should I kick the CI linux-focal-gcc-bazel-weekly-gurobi-debug?

**Review by SeanCurtis-TRI — COMMENTED** [review]
Let me rephrase my question. With snopt, gurobi, and mosek, we experienced timeouts on this optimization problem. Any idea why they timed out?

So, hard-coding to ipopt is a resolution of the *symptom* -- pick a solver we know doesn't time out. But does it explain why the symptom exists?

I'd recommend putting a comment at the site of the code change warning not to remove Ipopt for the symptom reason above. But I'm also wondering if we understand this well enough to know why the other solvers struggle. Is that something we should be worried about?

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14199#-:-MJSl7ij-KABD3Q8w-mT:b-5qaqe8)* status: needs platform reviewer assigned, needs at least two assigned reviewers

**Comment by hongkai-dai** [comment]
I believe the timeout is just due to SNOPT, should have nothing to do with Mosek and Gurobi.

I am not sure why SNOPT times out on ubuntu focal. On Ubuntu  bionic, we also tested with snopt, and that doesn't time out. My only guess is that some system libraries (like BLAS/LAPACK) changed from bionic to focal, and hence changed the performance of SNOPT.

Could you show me how to kick the CI linux-focal-gcc-bazel-weekly-gurobi-debug? If it passes the CI, then I will add a documentation in the code to say it is important to call Ipopt solver explicitly.

**Review by hongkai-dai — COMMENTED** [review]
+@seancurtis-tri for both reviews please, thanks!

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14199#-:-MJSxzQE5iOJ_QtSyLZ5:b79lg6w)* status: LGTM missing from assignee SeanCurtis-TRI(platform), needs at least two assigned reviewers (waiting on @SeanCurtis-TRI)

**Review by SeanCurtis-TRI — COMMENTED** [review]
Perhaps this is due to my ignorance, but I'm quite surprised by the assertion that the time out is solely attributable to SNOPT. 

This shows the failure with SNOPT. (The key here is `--confg=snopt` in the bazel configuration.)
https://drake-jenkins.csail.mit.edu/view/Weekly%20Production/job/linux-focal-clang-bazel-weekly-snopt-debug/19/

However, this shows the failure with gurobi (this time it's `--config=gurobi`.) 
https://drake-jenkins.csail.mit.edu/view/Weekly%20Production/job/linux-focal-clang-bazel-weekly-gurobi-debug/19/

And again for mosek
https://drake-jenkins.csail.mit.edu/view/Weekly%20Production/job/linux-focal-clang-bazel-weekly-mosek-debug/19/

Is it the case that SNOPT is a subset of Gurobi and Mosek? Is that where the assertion comes from?

As for testing this PR, you might try:

`linux-focal-clang-bazel-experimental-snopt-debug` as your CI target. (And if you swap "snopt" for "mosek" or "gurobi", there are experimental versions of those as well.)

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14199#-:-MJTuZyQ3hD_JWbxMIvp:b-udfaph)* status: LGTM missing from assignee SeanCurtis-TRI(platform), needs at least two assigned reviewers

**Comment by hongkai-dai** [comment]
@drake-jenkins-bot linux-focal-clang-bazel-experimental-snopt-debug please
@drake-jenkins-bot linux-focal-clang-bazel-experimental-debug please

**Comment by hongkai-dai** [comment]
@drake-jenkins-bot linux-focal-clang-bazel-experimental-snopt-debug please
@drake-jenkins-bot linux-focal-clang-bazel-experimental-mosek-debug please
@drake-jenkins-bot linux-focal-clang-bazel-experimental-gurobi-debug please
@drake-jenkins-bot linux-focal-clang-bazel-experimental-debug please

**Review by SeanCurtis-TRI — COMMENTED** [review]
(Had to look despite the no-review tag.)

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14199#-:-MJY8Z2u1QK_8t203VtT:b9di27m)* status: 2 unresolved discussions, LGTM missing from assignee SeanCurtis-TRI(platform), needs at least two assigned reviewers, labeled "do not merge" (waiting on @hongkai-dai and @SeanCurtis-TRI)

___
*[bindings/pydrake/multibody/BUILD.bazel, line 293 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14199#-MJY7b1R8iimv8CAk-0A:-MJY7b1SA1v3c9IhsxwF:b-lnvz5r) ([raw file](https://github.com/robotlocomotion/drake/blob/63798c3c2be8f240e51ab89b94aee818c5123e87/bindings/pydrake/multibody/BUILD.bazel#L293)):*
> ```Starlark
>     name = "optimization_test",
>     timeout = "moderate",
>     tags = ["snopt"],
> ```

BTW The implication of this is that this test no longer runs for non-snopt configurations (standard, mosek, or gurobi), right? So, you've nailed it down so that it works with snopt and have excluded other solvers.
___
*[bindings/pydrake/multibody/test/optimization_test.py, line 179 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14199#-MJY84Io2ct8VBnt3YXW:-MJY84Io2ct8VBnt3YXX:b-5hft8h) ([raw file](https://github.com/robotlocomotion/drake/blob/63798c3c2be8f240e51ab89b94aee818c5123e87/bindings/pydrake/multibody/test/optimization_test.py#L179)):*
> ```Python
>         self.assertTrue(result.is_success())
>         # Now progressively tighten the tolerance and solve it again.
>         for tol in (0.05, 0.01, 0.005, 0.001):
> ```

BTW Curious that for initial tolerance t, you use tolerance values: t, 0.5t, 0.1t, 0.05t, 0.01t. Kind of an interesting sampling of tolerances.  Equivalently, it looks like this sequence: t = 0.1, t /= 2, t /=5, t /=2, t /= 5. Seems kind of erratic. :)

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14199)* status: 2 unresolved discussions, LGTM missing from assignee SeanCurtis-TRI(platform), needs at least two assigned reviewers, labeled "do not merge" (waiting on @SeanCurtis-TRI)

___
*[bindings/pydrake/multibody/BUILD.bazel, line 293 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14199#-MJY7b1R8iimv8CAk-0A:-MJYW7CV-pNjOMLxXBt6:b-g1enmr) ([raw file](https://github.com/robotlocomotion/drake/blob/63798c3c2be8f240e51ab89b94aee818c5123e87/bindings/pydrake/multibody/BUILD.bazel#L293)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW The implication of this is that this test no longer runs for non-snopt configurations (standard, mosek, or gurobi), right? So, you've nailed it down so that it works with snopt and have excluded other solvers.
</blockquote></details>

That is right. In this optimization problem, we adopted a formulation that is recommended for SNOPT solver, as mentioned in Michael Posa's paper.
___
*[bindings/pydrake/multibody/test/optimization_test.py, line 179 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14199#-MJY84Io2ct8VBnt3YXW:-MJYWMwmAwJlGEzoBUj9:b-h2sm8j) ([raw file](https://github.com/robotlocomotion/drake/blob/63798c3c2be8f240e51ab89b94aee818c5123e87/bindings/pydrake/multibody/test/optimization_test.py#L179)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Curious that for initial tolerance t, you use tolerance values: t, 0.5t, 0.1t, 0.05t, 0.01t. Kind of an interesting sampling of tolerances.  Equivalently, it looks like this sequence: t = 0.1, t /= 2, t /=5, t /=2, t /= 5. Seems kind of erratic. :)
</blockquote></details>

Right. There is not too much scientific explanation on choosing this tolerance sequence. I tried this sequence and it works.

I also tried another sequence (0.1, 0.01, 0.001) and it failed. So I add some intermediate tolerance in between (0.1, 0.01 ) and (0.01, 0.001)

**Review by SeanCurtis-TRI — COMMENTED** [review]
:LGTM:
+@soonho-tri for platform review, please.

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14199#-:-MJc0wNN52Xg0jhYbBlB:bm8qn78)* status: LGTM missing from assignee soonho-tri(platform) (waiting on @soonho-tri)

**Review by soonho-tri — COMMENTED** [review]
:lgtm:

> Reviewed 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14199#-:-MJc6wQX4_teKHsqrtr4:bnfp4nl)* status: :shipit: complete! all discussions resolved, LGTM from assignees soonho-tri(platform),SeanCurtis-TRI(platform)
